primitives/storage/src/lib.rs

```diff
@@ -45,17 +45,56 @@ impl AsRef<[u8]> for StorageKey {
 #[cfg_attr(feature = "std", derive(Hash, PartialOrd, Ord))]
 pub struct TrackedStorageKey {
 	pub key: Vec<u8>,
-	pub has_been_read: bool,
-	pub has_been_written: bool,
+	pub reads: u32,
+	pub writes: u32,
+	pub whitelisted: bool,
 }
 
-// Easily convert a key to a `TrackedStorageKey` that has been read and written to.
+impl TrackedStorageKey {
+	/// Create a default `TrackedStorageKey`
+	pub fn new(key: Vec<u8>) -> Self {
+		Self {
+			key,
+			reads: 0,
+			writes: 0,
+			whitelisted: false,
+		}
+	}
+	/// Check if this key has been "read", i.e. it exists in the memory overlay.
+	///
+	/// Can be true if the key has been read, has been written to, or has been
+	/// whitelisted.
+	pub fn has_been_read(&self) -> bool {
+		self.whitelisted || self.reads > 0u32 || self.has_been_written()
+	}
+	/// Check if this key has been "written", i.e. a new value will be committed to the database.
+	///
+	/// Can be true if the key has been written to, or has been whitelisted.
+	pub fn has_been_written(&self) -> bool {
+		self.whitelisted || self.writes > 0u32
+	}
+	/// Add a storage read to this key.
+	pub fn add_read(&mut self) {
```

Comment:
nit: if you have getter and setters, the fields better be private. With public fields, most of these methods are considered redundant.

Reply:
these functions are just for convenience and ergonomics